### PR TITLE
Fix for Issue #27 VHD file contains more than one partition

### PR DESCRIPTION
**Pull request by TheBlackMini**
Added an extra non-mandatory parameter to allow users to specify partition number.
Added localization.
Re-formatted the code (OCD kicked-in)

---

This change is [Reviewable](https://reviewable.io/reviews/powershell/xhyper-v/59)

### CODE REVIEW

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 [hunk 1]
```diff
@@ -1,8 +1,39 @@
+# Fallback message strings in en-US
+DATA localizedData
+{
+    # same as culture = "en-US"
+ConvertFrom-StringData @'    
+    IncorrectPartitionCount=Partitions on VHD has {0] partitions. Unable to continue without specifying partition numbe
```
Comment (FayzaSultan) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: >numbe [](start = 114, length = 5)

Spelling. Number

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof [hunk 2]
```diff
@@ -1,23 +1,10 @@
...
-};
-
-[ClassVersion("1.0.0.0"), FriendlyName("xVhdFile")]
+[ClassVersion("1.0.0.1"), FriendlyName("xVhdFile")]
 class MSFT_xVhdFileDirectory : OMI_BaseResource
 {
     [Key, Description("Path to the VHD")] String VhdPath;
-    [Required, EmbeddedInstance("MSFT_xFileDirectory"), Description("The FileDirectory objects to copy to the VHD")] String FileDirectory[];
-        [Write,ValueMap{"ModifiedDate","SHA-1","SHA-256","SHA-512"},Values{"ModifiedDate","SHA-1","SHA-256","SHA-512"}] string CheckSum;
+    [Required, EmbeddedInstance("MSFT_FileDirectoryConfiguration"), Description("The FileDirectory objects to copy to the VHD")] String FileDirectory[];
```
Comment (FayzaSultan) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof]: >[Required, EmbeddedInstance("MSFT_FileDirectoryConfiguration"), Description("The FileDirectory objects to copy to the VHD")] String FileDirectory[]; [](start = 4, length = 148)

Can you update the examples here xHyper-V/Examples/Sample_xVhdFileExamples.ps1  to use MSFT_FileDirectoryConfiguration ?

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof [hunk 3]
```diff
@@ -1,23 +1,10 @@
...
+    [Write, Description("Specify partition number if more than one partition is present in the VHD")] Uint32 VhdPartitionNumber;
```
Comment (FayzaSultan) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof]: >   [Write, Description("Specify partition number if more than one partition is present in the VHD")] Uint32 VhdPartitionNumber; [](start = 1, length = 127)

Can we add some tests that use this new parameter ?

**Comment by FayzaSultan** [comment]
> ### Unreleased

Need to add information about new change in unreleased notes.

---
Refers to: README.md:80 in 3b5f7fa. [](commit_id = 3b5f7faa434c0643c96c3e29aa21ca8baaf9a6f4, deletion_comment = False)

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof (hunk 2 above)
Comment (TheBlackMini) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof]: Created a new example file.

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof (hunk 3 above)
Comment (TheBlackMini) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof]: Not sure what you mean? Could you elaborate?

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 [hunk 4]
```diff
@@ -22,39 +57,48 @@ function Get-TargetResource
...
-        $letterDrive  = (-join $mountedDrive.DriveLetter) + ":\"
-       
-        # show the drive letters.
-        Get-PSDrive | Write-Verbose       
-
-        $finalPath = Join-Path $letterDrive $item.DestinationPath
+    }
+    
+    # Turn drive letter into a path
+    $letterDrive  = "$($mountedDrive.DriveLetter):\"
```
Comment (eshaparmar) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: Error handling for when $mountedDrive is $null

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 [hunk 5]
```diff
@@ -88,62 +137,73 @@ function Set-TargetResource
...
+        # Retrieve the current drive letter for the VHD
+        if($VhdPartitionNumber)
+        {
+            $mountedDrive =  $mountVHD | Get-Disk | Get-Partition | Where-Object -FilterScript { $_.PartitionNumber -eq $VhdPartitionNumber } | Get-Volume
+        }
+        else
+        {
+            $mountedDrive =  $mountVHD | Get-Disk | Get-Partition | Get-Volume
+        }
+
```
Comment (eshaparmar) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: Can mounteddrive be null? If so, please add error handling for that

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 [hunk 6]
```diff
@@ -88,62 +137,73 @@ function Set-TargetResource
...
+    $mountVHD = EnsureVHDState -Mounted -vhdPath $VhdPath
 
     try
     {
-            # show the drive letters.
-            Get-PSDrive | Write-Verbose
-
-            $mountedDrive = $mountedVHD | Get-Disk | Get-Partition | Get-Volume
+        # Retrieve the current drive letter for the VHD
+        if($VhdPartitionNumber)
```
Comment (eshaparmar) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: Looks like there is some common code that can be put in a separate function and shared across test/get/set. For e.g. getting $mountedDrive

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 [hunk 7]
```diff
@@ -172,70 +236,88 @@ function Test-TargetResource
...
     # mount the vhd.
-    $mountedVHD = EnsureVHDState -Mounted -vhdPath $VhdPath
+    $mountVHD = EnsureVHDState -Mounted -vhdPath $VhdPath
 
     try 
     {
-        # Show the drive letters after mount 
-        Get-PSDrive | Write-Verbose
+        # Retrieve the current drive letter for the VHD
+        if($VhdPartitionNumber)
```
Comment (eshaparmar) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: Common code. Same comment as before

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 [hunk 8]
```diff
@@ -365,74 +450,107 @@ function GetItemToCopy
...
-
     }
-    elseif ($PSCmdlet.ParameterSetName -eq 'Set')
+    elseif($PSCmdlet.ParameterSetName -eq 'Set')
     {
-        Write-Verbose "Attempting to change the attribute of the file $destinationPath to value $attribute"
-        Set-ItemProperty -Path $destinationPath -Name Attributes -Value $attribute
+        try
+        {
+            Set-ItemProperty -Path $destinationPath -Name Attributes -Value $attribute -ErrorAction Stop
```
Comment (eshaparmar) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: We writing errors to the verbose stream. Shouldn't these be terminating?

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof (hunk 3 above)
Comment (eshaparmar) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof]: Please add new unit tests and integration tests for the new parameter - VhdPartitionNumber

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 (hunk 8 above)
Comment (TheBlackMini) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: These aren't error logs, just states that it worked if it goes past the Set-ItemProperty

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof (hunk 3 above)
Comment (TheBlackMini) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof]: The Uint32 is the validation for the parameter.

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 (hunk 4 above)
Comment (TheBlackMini) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: New function which handles this now.

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 (hunk 6 above)
Comment (TheBlackMini) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: New function which handles this now.

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 (hunk 5 above)
Comment (TheBlackMini) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: New function which handles this now.

File: DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1 (hunk 7 above)
Comment (TheBlackMini) [inline DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1]: New function which handles this now.

**Comment by TheBlackMini** [comment]
Updated code

**Comment by TheBlackMini** [comment]
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1, line 6 at r1](https://reviewable.io:443/reviews/powershell/xhyper-v/59#-KQU93STgbJXrSkSZR1C-r1-6:-KSJQwIdARxQnqVrPeSi:bayklsb) ([raw file](https://github.com/powershell/xhyper-v/blob/3b5f7faa434c0643c96c3e29aa21ca8baaf9a6f4/DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1#L6)):*
<details><summary><i>Previously, FayzaSultan wrote…</i></summary>
> >numbe [](start = 114, length = 5)
> 
> 
> 
> Spelling. Number 
</details>
Fixed spelling

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/59)*

**Comment by TheBlackMini** [comment]
Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1, line 6 at r1](https://reviewable.io:443/reviews/powershell/xhyper-v/59#-KQU93STgbJXrSkSZR1C-r1-6:-KSJRHDkjzx1EBl0299B:b-896fix) ([raw file](https://github.com/powershell/xhyper-v/blob/3b5f7faa434c0643c96c3e29aa21ca8baaf9a6f4/DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1#L6)):*
<details><summary><i>Previously, TheBlackMini (Adam) wrote…</i></summary>
> Fixed spelling
</details>
Done.

---

*[DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1, line 85 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/59#-KQU93STgbJXrSkSZR1C-r2-85:-KSJRBDjeP3AXxoVa-yD:b-896fix) ([raw file](https://github.com/powershell/xhyper-v/blob/1b04b4fa8476ac5cd5abe523a07f1638d98d6fe5/DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1#L85)):*
<details><summary><i>Previously, TheBlackMini (Adam) wrote…</i></summary>
> New function which handles this now.
</details>
Done.

---

*[DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1, line 152 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/59#-KQU93STgbJXrSkSZR1C-r2-152:-KSJRC1pNHg0Wd2UR4ad:b-896fix) ([raw file](https://github.com/powershell/xhyper-v/blob/1b04b4fa8476ac5cd5abe523a07f1638d98d6fe5/DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1#L152)):*
<details><summary><i>Previously, TheBlackMini (Adam) wrote…</i></summary>
> New function which handles this now.
</details>
Done.

---

*[DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1, line 160 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/59#-KQU93STgbJXrSkSZR1C-r2-160:-KSJRCPlTMTo52yrSKK9:b-896fix) ([raw file](https://github.com/powershell/xhyper-v/blob/1b04b4fa8476ac5cd5abe523a07f1638d98d6fe5/DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1#L160)):*
<details><summary><i>Previously, TheBlackMini (Adam) wrote…</i></summary>
> New function which handles this now.
</details>
Done.

---

*[DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1, line 251 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/59#-KQU93STgbJXrSkSZR1C-r2-251:-KSJRCodY_vWo5zbdyxd:b-896fix) ([raw file](https://github.com/powershell/xhyper-v/blob/1b04b4fa8476ac5cd5abe523a07f1638d98d6fe5/DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1#L251)):*
<details><summary><i>Previously, TheBlackMini (Adam) wrote…</i></summary>
> New function which handles this now.
</details>
Done.

---

*[DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1, line 509 at r2](https://reviewable.io:443/reviews/powershell/xhyper-v/59#-KQU93STgbJXrSkSZR1C-r2-509:-KSJRF5G8NYApT2QML6K:b-896fix) ([raw file](https://github.com/powershell/xhyper-v/blob/1b04b4fa8476ac5cd5abe523a07f1638d98d6fe5/DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.psm1#L509)):*
<details><summary><i>Previously, TheBlackMini (Adam) wrote…</i></summary>
> These aren't error logs, just states that it worked if it goes past the Set-ItemProperty
</details>
Done.

---

*[DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof, line 5 at r1](https://reviewable.io:443/reviews/powershell/xhyper-v/59#-KQU93STgbJXrSkSZR1D-r1-5:-KSJRF_tF1ROLxuDyfpT:b-896fix) ([raw file](https://github.com/powershell/xhyper-v/blob/3b5f7faa434c0643c96c3e29aa21ca8baaf9a6f4/DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof#L5)):*
<details><summary><i>Previously, TheBlackMini (Adam) wrote…</i></summary>
> Created a new example file.
</details>
Done.

---

*[DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof, line 6 at r1](https://reviewable.io:443/reviews/powershell/xhyper-v/59#-KQU93STgbJXrSkSZR1D-r1-6:-KSJRFzcZ8POSPYSBr0m:b-896fix) ([raw file](https://github.com/powershell/xhyper-v/blob/3b5f7faa434c0643c96c3e29aa21ca8baaf9a6f4/DSCResources/MSFT_xVhdFileDirectory/MSFT_xVhdFileDirectory.schema.mof#L6)):*
<details><summary><i>Previously, TheBlackMini (Adam) wrote…</i></summary>
> The Uint32 is the validation for the parameter.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xhyper-v/59)*

**Comment by kwirkykat** [comment]
@TheBlackMini Looks like your tests are failing because there is a tab indentation somewhere instead of spaces. Can you please fix it so that the tests pass in AppVeyor?

**Comment by bgelens** [comment]
@TheBlackMini Since this PR has been stalled for some time now, I'm assigning the abandoned label. Please let me know if you still want to work on this or not.

**Pull request closed without merging**